examples/typescript-plugin-safety.ts

```diff
@@ -0,0 +1,99 @@
+import { FastifyDecorators, FastifyInstance } from "../types/instance";
+
+type ExtractSixGeneric<T> = T extends FastifyInstance<any, any, any, any, any, infer U> ? U : void;
+type ExtractFirstFunctionParameter<T> = T extends (...args: infer P) => any ? P[0] : void;
+export type UnionToIntersection<U> = (
+  U extends unknown ? (k: U) => void : never
+) extends (k: infer I) => void
+  ? I
+  : never;
+
+// Plugin definition on the `fastify-plugin` library
+type PluginMetadata = {
+    /** Bare-minimum version of Fastify for your plugin, just add the semver range that you need. */
+    fastify?: string,
+    name?: string,
+    /** Decorator dependencies for this plugin */
+    decorators?: {
+        fastify?: (string | symbol)[],
+        reply?: (string | symbol)[],
+        request?: (string | symbol)[]
+    },
+    /** The plugin dependencies */
+    dependencies?: string[],
+    encapsulate?: boolean
+}
+
+type FastifyPluginDecorators = {
+    decorators: FastifyDecorators,
+    dependencies: (FastifyPluginCallback<any, any, any, any, any> | FastifyPluginAsync<any, any, any, any, any>)[]
+}
+
+declare function fastifyPlugin<
+    Decorators extends FastifyPluginDecorators = { decorators: {}, dependencies: [] },
+    Options extends FastifyPluginOptions = Record<never, never>,
+    RawServer extends RawServerBase = RawServerDefault,
+    TypeProvider extends FastifyTypeProvider = FastifyTypeProviderDefault,
+    Logger extends FastifyBaseLogger = FastifyBaseLogger,
+    Fn extends FastifyPluginCallback<Options, RawServer, TypeProvider, Logger,  Decorators["decorators"]> | FastifyPluginAsync<Options, RawServer, TypeProvider, Logger, Decorators["decorators"]> = FastifyPluginCallback<Options, RawServer, TypeProvider, Logger,  Decorators["decorators"]>
+>(
+    fn: Fn extends unknown ? Fn extends (...args: any) => Promise<any> ? FastifyPluginAsync<
+        Options, RawServer, TypeProvider, Logger, Decorators["decorators"] & ExtractSixGeneric<ExtractFirstFunctionParameter<Decorators["dependencies"][number] extends undefined ? {} : Decorators["dependencies"][number]>>
+    > : FastifyPluginCallback<
+        Options, RawServer, TypeProvider, Logger, Decorators["decorators"] & ExtractSixGeneric<ExtractFirstFunctionParameter<Decorators["dependencies"][number] extends undefined ? {} : Decorators["dependencies"][number]>>
+    > : Fn,
+    options?: PluginMetadata | string,
+): Fn;
+
+// Used to type plugins written in javascript
+export type GetPluginTypes<Decorators extends FastifyPluginDecorators = { decorators: {}, dependencies: [] }> = FastifyPluginCallback<{}, RawServerBase, FastifyTypeProviderDefault, FastifyBaseLogger,  Decorators["decorators"] & ExtractSixGeneric<ExtractFirstFunctionParameter<Decorators["dependencies"][number] extends undefined ? {} : Decorators["dependencies"][number]>>>;
+
+// Testing types
```

Comment:
Typescript tests should live in `test/types` folder 😉

Reply:
Actually `GetPluginTypes` should not be living on the test folder at all, I just put it there for demo purposes. I think I should move it on the type definitions, since it will be exported and used by the libraries written in javascript (no library should be written in javascript IMO) or they can still use the declaration merging we currently have